### PR TITLE
chore(observability): count byte_size after transforming event

### PR DESCRIPTION
**Pull request by StephenWakely**
Before this the `component_sent_event_bytes_total` was emitting the json byte size of the event including any fields dropped by `only_fields` and `except_field` settings.

This changes it so the the count is made after transforming the event. 

A complication arose whereby if the service field was dropped we would no longer have access to this value to emit the `service` tag with the metrics. This also adds a `dropped_fields` field to the event metadata where this value can be stored and accessed.

### CODE REVIEW

File: lib/vector-core/src/event/metadata.rs [hunk 1]
```diff
@@ -123,6 +131,17 @@ impl EventMetadata {
     pub fn set_splunk_hec_token(&mut self, secret: Arc<str>) {
         self.secrets.insert(SPLUNK_HEC_TOKEN, secret);
     }
+
+    /// Adds the value to the dropped fields list, returning the
+    /// position the value will be found in the list.
```
Comment (spencergilbert) [inline lib/vector-core/src/event/metadata.rs:136]: Is this comment correct? The function doesn't have a return.

File: src/codecs/encoding/transformer.rs [hunk 2]
```diff
@@ -128,11 +128,23 @@ impl Transformer {
...
         if let Some(only_fields) = self.only_fields.as_ref() {
-            let old_value = std::mem::replace(log.value_mut(), Value::Object(BTreeMap::new()));
+            let mut old_value = std::mem::replace(log.value_mut(), Value::Object(BTreeMap::new()));
 
             for field in only_fields {
-                if let Some(value) = old_value.get(field) {
-                    log.insert((PathPrefix::Event, field), value.clone());
+                if let Some(value) = old_value.remove(field, true) {
+                    log.insert((PathPrefix::Event, field), value);
+                }
```
Comment (spencergilbert) [inline src/codecs/encoding/transformer.rs:136]: Is this get -> remove change just an optimization, it doesn't seem related to the functional change?

File: src/codecs/encoding/transformer.rs [hunk 3]
```diff
@@ -141,7 +153,20 @@ impl Transformer {
...
         if let Some(except_fields) = self.except_fields.as_ref() {
             for field in except_fields {
-                log.remove(field.as_str());
+                let value = log.remove(field.as_str());
+
+                // If we are removing the service field we need to store this in a `dropped_fields` list as we may need to
+                // refer to this later when emitting metrics.
+                if let Some(v) = value {
+                    if let Some(path) = log.metadata().schema_definition().meaning_path("service") {
+                        use lookup::path::TargetPath;
```
Comment (spencergilbert) [inline src/codecs/encoding/transformer.rs:162]: I think it's unusual seeing `use` statements this deep in the code.

File: src/sinks/new_relic/encoding.rs [hunk 4]
```diff
@@ -22,7 +24,11 @@ impl Encoder<Result<NewRelicApiModel, NewRelicSinkError>> for NewRelicEncoder {
             writer.write_all(json)?;
             Ok(())
         })?;
-        io::Result::Ok(size)
+
+        // TODO This should not be zero.
```
Comment (spencergilbert) [inline src/sinks/new_relic/encoding.rs:28]: Handle before merging?

File: src/sinks/new_relic/encoding.rs [hunk 5]
```diff
@@ -22,7 +24,11 @@ impl Encoder<Result<NewRelicApiModel, NewRelicSinkError>> for NewRelicEncoder {
...
+        let byte_size = telemetry().create_request_count_byte_size();
```
Comment (dsmith3197) [inline src/sinks/new_relic/encoding.rs:29]: What's the impact of this, and do we have any plan to resolve it?

File: lib/vector-core/src/event/metadata.rs (hunk 1 above)
Comment (StephenWakely) [reply lib/vector-core/src/event/metadata.rs:136]: Good catch!

File: src/codecs/encoding/transformer.rs (hunk 2 above)
Comment (StephenWakely) [reply src/codecs/encoding/transformer.rs:136]: 1. It removes the need to `clone` the value when inserting into the new `log`.
2. By removing it from `old_value` it indicates that the field has been added to the log, so below when we are getting the value meaning `service` it won't exist in `old_value` and thus we don't need to add it to `dropped_fields`.

It's a beautiful part of Rust, because there is no `clone` this field can only exist at most once - either in the log message or in the `dropped_fields`.

File: src/codecs/encoding/transformer.rs (hunk 3 above)
Comment (StephenWakely) [reply src/codecs/encoding/transformer.rs:162]: Hmm.. I'm not sure if I want that trait to be available in the whole module because it adds function to `&str`. So I am trying to limit the scope to just here.

Comment (StephenWakely) [reply src/codecs/encoding/transformer.rs]: I've reorganised the function a bit.

File: src/sinks/new_relic/encoding.rs (hunk 4 above)
Comment (StephenWakely) [reply src/sinks/new_relic/encoding.rs]: Yes. I've had to restructure the `new_relic` sink to move more functionality into the `Encoder` - I think it's more correct now and we can get the sizes correctly.

There are no integration tests for new relic so I haven't been able to test this fully, so a thorough check would be appreciated.

File: src/sinks/new_relic/encoding.rs (hunk 5 above)
Comment (StephenWakely) [reply src/sinks/new_relic/encoding.rs]: Yes. I've had to restructure the `new_relic` sink to move more functionality into the `Encoder` - I think it's more correct now and we can get the sizes correctly.

There are no integration tests for new relic so I haven't been able to test this fully, so a thorough check would be appreciated.

File: src/codecs/encoding/transformer.rs (hunk 3 above)
Comment (spencergilbert) [reply src/codecs/encoding/transformer.rs]: Got it, just making sure it was intentional - I've had IDE's sometimes stick those in more willy-nilly than I'd like.

File: src/sinks/new_relic/config.rs [hunk 6]
```diff
@@ -165,7 +155,10 @@ impl SinkConfig for NewRelicConfig {
         let sink = NewRelicSink {
             service,
             transformer: self.encoding.clone(),
```
Comment (dsmith3197) [inline src/sinks/new_relic/config.rs:157]: It looks like we can remove `transformer` from the top-level struct now.
```suggestion
```

**Review by dsmith3197 — DISMISSED** [review]
lgtm

File: lib/vector-core/src/event/log_event.rs [hunk 7]
```diff
@@ -293,10 +293,14 @@ impl LogEvent {
...
-            .schema_definition()
-            .meaning_path(meaning.as_ref())
-            .and_then(|path| self.get(path))
+        if let Some(dropped) = self.metadata().dropped_field(&meaning) {
+            Some(dropped)
+        } else {
+            self.metadata()
+                .schema_definition()
+                .meaning_path(meaning.as_ref())
+                .and_then(|path| self.get(path))
```
Comment (lukesteensen) [inline lib/vector-core/src/event/log_event.rs:302]: Is it possible for these two paths to get out of sync? E.g. if you drop a field with a given meaning and then subsequently write a new field and assign it that same meaning? Or is that not a problem because we're only using this in the `{only,except}_fields` path?

File: src/sinks/amqp/encoder.rs [hunk 8]
```diff
@@ -11,9 +12,17 @@ pub(super) struct AmqpEncoder {
...
+    ) -> io::Result<(usize, GroupedCountByteSize)> {
         let mut body = BytesMut::new();
         self.transformer.transform(&mut input);
+
+        let mut byte_size = telemetry().create_request_count_byte_size();
+        byte_size.add_event(&input, input.estimated_json_encoded_size_of());
+
         let mut encoder = self.encoder.clone();
         encoder
             .encode(input, &mut body)
```
Comment (lukesteensen) [inline src/sinks/amqp/encoder.rs:28]: This is a bit outside the scope of the PR, but I'm curious why we need to estimate the JSON size right before we actually encode the event. Don't we have the actual size here that we could use directly?

Comment (jszwedko) [reply src/sinks/amqp/encoder.rs:28]: This was part of https://github.com/vectordotdev/vector/issues/17576. The intention was to use a measurement that could be taken anywhere in the processing pipeline (source, transforms, sinks) so that the values are comparable.

**Review by dsmith3197 — APPROVED** [review]

File: lib/vector-core/src/event/log_event.rs (hunk 7 above)
Comment (lukesteensen) [reply lib/vector-core/src/event/log_event.rs:306]: If we do end up going with some kind of change like #18028 then it would be nice to undo this just to get rid of the oddity in the API and potential for mismatch. I'm not sure if that's just something like a TODO on the field or an issue, but maybe worth noting so that this doesn't stick around in the event that it becomes unneeded.

**Review by lukesteensen — APPROVED** [review]
